File: app/src/main/java/org/cerion/stockcharts/ui/symbols/SymbolsFragment.java
package org.cerion.stockcharts.ui.symbols;

import android.app.DialogFragment;
import android.app.Fragment;
import android.app.FragmentTransaction;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.cerion.stockcharts.R;
import org.cerion.stockcharts.common.SymbolLookupDialogFragment;
import org.cerion.stockcharts.common.ViewModelFragment;


public class SymbolsFragment extends ViewModelFragment<SymbolsViewModel> implements SymbolLookupDialogFragment.OnSymbolListener {

    private static final String TAG = SymbolsFragment.class.getSimpleName();

    @Override
    protected SymbolsViewModel newViewModel() {
        return new SymbolsViewModel(getActivity());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_symbols, container, false);

        RecyclerViewAdapter adapter = new RecyclerViewAdapter(getViewModel().items);

        RecyclerView rv = view.findViewById(R.id.recycler_view);
        rv.setHasFixedSize(true);
        rv.setLayoutManager(new LinearLayoutManager(getActivity()));
        rv.setAdapter(adapter);
        setHasOptionsMenu(true);
        return view;
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        if (item.getGroupId() == RecyclerViewAdapter.CONTEXT_MENU_GROUP_ID) {
            SymbolItemViewModel vm = getViewModel().items.get(item.getOrder());

            switch (item.getItemId()) {
                case RecyclerViewAdapter.CONTEXT_MENU_DELETE:
                    getViewModel().delete(vm.getSymbol());
                    break;
                default:
                    return super.onContextItemSelected(item);
            }

            return true;
        }

        return false;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        getViewModel().load();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.symbol_list_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch(item.getItemId()) {
            case R.id.add_symbol:
                onAddSymbol();
                break;
            default:
                return super.onContextItemSelected(item);
        }

        return true;
    }

    private static final int DIALOG_FRAGMENT = 1;
    private void onAddSymbol() {
        //DialogFragment newFragment = new SymbolLookupDialogFragment();
        //newFragment.show(getFragmentManager(), "addSymbol");

        // TODO simpler way of doing this?
        FragmentTransaction ft = getActivity().getFragmentManager().beginTransaction();
        Fragment prev = getActivity().getFragmentManager().findFragmentByTag("dialog");
        if (prev != null) {
            ft.remove(prev);
        }
        ft.addToBackStack(null);

        DialogFragment dialog = new SymbolLookupDialogFragment();
        dialog.setTargetFragment(this, DIALOG_FRAGMENT);
        dialog.show(getFragmentManager().beginTransaction(), "dialog");
    }

    @Override
    public void onSymbolEntered(final String name) {
        getViewModel().add(name);
    }
}

File: app/src/main/java/org/cerion/stockcharts/ui/symbols/SymbolsFragment.kt
package org.cerion.stockcharts.ui.symbols

import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.cerion.stockcharts.R
import org.cerion.stockcharts.ui.HomeFragmentDirections
import org.koin.androidx.viewmodel.ext.android.viewModel

class SymbolsFragment : Fragment() {

    private val viewModel: SymbolsViewModel by viewModel()

    companion object {
        fun newInstance(category: SymbolCategory): SymbolsFragment {
            val fragment = SymbolsFragment()
            val args = Bundle()
            args.putString("category", category.name)
            fragment.arguments = args
            return fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.fragment_symbols, container, false)

        val adapter = RecyclerViewAdapter(object : RecyclerViewAdapter.SymbolListener {
            override fun click(symbol: String) {
                val action = HomeFragmentDirections.actionFragmentHomeToChartsFragment(symbol)
                findNavController().navigate(action)
            }
        })

        val rv: RecyclerView = view.findViewById(R.id.recycler_view)
        rv.setHasFixedSize(true)
        rv.layoutManager = LinearLayoutManager(activity)
        rv.adapter = adapter
        setHasOptionsMenu(true)

        viewModel.items.observe(viewLifecycleOwner, Observer {
            adapter.setItems(it)
        })

        arguments?.getString("category")?.also {
            viewModel.load(SymbolCategory.valueOf(it))
        }

        return view
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        if (item.groupId == RecyclerViewAdapter.CONTEXT_MENU_GROUP_ID) {

            val s = viewModel.items.value!!.keys.toList()[item.order]
            when (item.itemId) {
                RecyclerViewAdapter.CONTEXT_MENU_DELETE -> viewModel.delete(s)
                else -> return super.onContextItemSelected(item)
            }
            return true
        }
        return false
    }

    @Deprecated("Deprecated in Java")
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.symbol_list_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    @Deprecated("Deprecated in Java")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_symbol -> onAddSymbol()
            else -> return super.onContextItemSelected(item)
        }

        return true
    }

    private fun onAddSymbol() { //DialogFragment newFragment = new SymbolLookupDialogFragment();
        val fm = requireActivity().supportFragmentManager
        val dialog = SymbolLookupDialog()
        dialog.show(fm, SymbolLookupDialog::class.java.simpleName)
    }
}